src/lightning/fabric/utilities/cloud_io.py

```diff
@@ -66,7 +65,5 @@ def _atomic_save(checkpoint: Dict[str, Any], filepath: Union[str, Path]) -> None
         filepath: The path to which the checkpoint will be saved.
             This points to the file that the checkpoint will be stored in.
     """
-    bytesbuffer = io.BytesIO()
```

Comment:
I took a dive into the project history.

Somebody had a problem with ordinary saving failing and corrupting the checkpoints, so an intermediate save was added: https://github.com/Lightning-AI/lightning/pull/689

But this created unnecessary remote IO, which prompted the buffer change: https://github.com/Lightning-AI/lightning/pull/2925

This means that doing this proposed change might reintroduce corrupted checkpoints. It's possible that torch has improved in this regard and it's not a problem anymore but this would need to be investigated.

Reply:
I'm not sure writing files is able to survive SIGKILL in the first place. Dumping the model into a bytes buffer first and then write could still give you partial writes of new checkpoint, i.e. `f.write(bytesbuffer.getvalue())` is not atomic with respect to the remote file system.

For example, after successfully writing checkpoint v1, the trainer is killed before it completes writing checkpoint v2, then the model is still corrupted.

IMHO the only thing we could promise to our user is that v1 should not be corrupted w/ potential partial writes of v2.  and This could be achieved by writing a new temporary path and then rename it into the old path, instead of in-place update of old checkpoint.

Reply:
> IMHO the only thing we could promise to our user is that v1 should not be corrupted w/ potential partial writes of v2. This could be achieved by writing a temporary new path and then rename it into the old path, instead of in-place update of old checkpoint.

If we do this, we are effectively reverting #2925 whose intention was to avoid writing to an intermediate path because of slowness with remote filesystems.

This change would not be a big deal for the `ModelCheckpoint`, because it uses a versioning scheme by default and only deletes after the new one was saved successfully. But we still need to consider that this `_atomic_save` function is called by all components.

Looking in the wild, we can see both options used:
https://github.com/pytorch/torchsnapshot/blob/087caae16ac83c99783fdb698fa1e3804a2a1670/torchsnapshot/serialization.py#L247-L250
https://github.com/microsoft/DeepSpeed/blob/195563a2c5dd865c2f56df133e14249c1387171c/deepspeed/runtime/checkpoint_engine/torch_checkpoint_engine.py#L22

Reply:
AFAIK `_atomic_save` is only used in saving model checkpoints: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/io/torch_io.py#L57-L65

Please correct me if I am wrong, e.g. there are other uses of this utility function.

Reply:
> writing to an intermediate path because of slowness with remote filesystems.

I don't get this. Eventually you will need to write to remote filesystem once (and better only once).

Does it write checkpoint twice to remote path in #2925?

Reply:
Before #2925 it wrote twice (temporary and final) and after it, it wrote to the buffer which seems to have the memory issue you are reporting.

Reply:
Then IMHO it really makes no difference whether dumping to a buffer or not before writing it once and only once to the remote filesystem.